### PR TITLE
fix: resolve output artifact of steps from expression when it refers …

### PR DESCRIPTION
**Pull request by shuangkun**
Solve the question mentioned:
https://github.com/argoproj/argo-workflows/issues/9120

### CODE REVIEW

File: workflow/controller/scope.go [hunk 1]
```diff
@@ -103,7 +103,12 @@ func (s *wfScope) resolveArtifact(art *wfv1.Artifact) (*wfv1.Artifact, error) {
 	}
 	valArt, ok := val.(wfv1.Artifact)
 	if !ok {
-		return nil, errors.Errorf(errors.CodeBadRequest, "Variable {{%v}} is not an artifact", art)
+		ptArt, ok := val.(*wfv1.Artifact)
```
Comment (terrytangyuan) [inline workflow/controller/scope.go:106]: Could you explain why this is needed? Seems redundant to the above lines?

Comment (shuangkun) [reply workflow/controller/scope.go:106]: Like the workflow mentioned，when i submit the workflow which refers itself input artfacts in fromExpression,  it return a pointer (*wfv1.Artifact)， so i try to transform it to a point when it transform to wfv1.Artifact failed.

kind: Workflow
metadata:
  generateName: from-expression-bug-
spec:
  entrypoint: main
  arguments:
    artifacts:
    - name: foo
      raw:
        data: |
          Hello
  templates:
    - name: main
      inputs:
        artifacts:
        - name: foo
      steps:
      - - name: hello
          inline:
            container:
              image: alpine:latest
      outputs:
        artifacts:
        - name: result
          # a trivial expression here only for reproducing the bug
          fromExpression: "1 == 1 ? inputs.artifacts.foo : inputs.artifacts.foo"

**Review by terrytangyuan — COMMENTED** [review]
Can you add a comment in the code as well as a test case?

**Comment by shuangkun** [comment]
> Can you add a comment in the code as well as a test case?

I have add a unit test and comment.
Before I add the pr, there is a error:
image
After add the code, the ut pass:

**Review by terrytangyuan — APPROVED** [review]
Thank you!

**Pull request merged**